### PR TITLE
Add Publisher.flatMapConcatSingle

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Publisher.flatMapConcatSingle provides similar semantics to
Publisher.flatMapMergeSingle except the results are returned
in the same order as the original Publisher.

### CODE REVIEW

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/1716

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java
Comment (Scottmitch) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:40]: note there is a queue here, and in `flatMapMergeSingle`. we can optimize and combine in the future with a native operator implementation but for now reusing `flatMapMergeSingle` is good enough to get started and keeps complexity down.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java
Comment (bondolo) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:894]: The asynchronous and unordered characteristic should be covered in the first sentence to distinguish this operation from `map()`

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java
Comment (bondolo) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:1333]: Perhaps contrast with `flatMapMergeSingle` or `@see flatMapMergeSingle`

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java
Comment (bondolo) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:44]: perhaps just call `flatmapConcat(publisher, mapper, 4)` except for the lower bounds issue…

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java
Comment (bondolo) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:61]: this lower bounds is not mentioned in the Publisher docs.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java
Comment (bondolo) [inline servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:90]: Perhaps`get()` → `intValue()`. When I see `get()` I always think "reference".

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:991]: Because merge is unordered, consider using [CompletableFuture.allOf(...)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#allOf-java.util.concurrent.CompletableFuture...-) to demonstrate that.

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatSingleTest.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatSingleTest.java:51]: Consider:
```
@ParameterizedTest(name = "{displayName} [{index}] begin={0} end={1} maxConcurrency={2} delayError={3}")
```

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatSingleTest.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatSingleTest.java:55]: For all tests: to make sure each `Single` completes at different time in different order, consider using a random delay <=5ms. Currently, there is a risk that ordering is guaranteed by having the same timer delay.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:88]: If we use named `Runnable`(s) (static inner classes) in `PublisherFlatMapConcatUtils` we can reuse existing utilities that operate with `AtomicIntegerFieldUpdater` and save some memory.

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatSingleTest.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatSingleTest.java:36]: Should not block this PR, but will be nice to have a test that verifies AsyncContext is propagated correctly to make sure all `shareContextOnSubscribe()` operators placed correctly.

**Review by idelpivnitskiy — APPROVED** [review]
LGTM after comments :)

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:44]: The goal is to inherit the limits/bounds/default from `flatMapMergeSingle` unless explicitly specified.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapConcatUtils.java:61]: The queue is unbound, the parameter is just an estimate for initial size of the queue. This value is not appropriate for javadocs as it is not part of the API contract and can be adjusted if necessary.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:991]: `delayError` error propagation is different than `CompletableFuture.allOf(..)` because it may potentially return all exceptions (e.g. not bail on the first one).

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java:88]: good catch! named `Function` will do the trick in this case based upon where state lives.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:1333]: I'll add a `@see flatMapMergeSingle` for now to keep javadocs compressed.

File: servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java
Comment (Scottmitch) [reply servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java]: I'll just leave it as `get()` bcz that is the accessor for `Atomic*` types

**Pull request merged**
